File: 2296.design-a-text-editor.py
#
# @lc app=leetcode id=2296 lang=python3
#
# [2296] Design a Text Editor
#

# @lc code=start
class TextEditor:

    def __init__(self):
        # Stores characters to the left of the cursor
        self.left_stack = []
        # Stores characters to the right of the cursor in reverse order
        # Example: if text is "abc|def", left=['a','b','c'], right=['f','e','d']
        self.right_stack = []

    def addText(self, text: str) -> None:
        # Add characters to the left stack
        for char in text:
            self.left_stack.append(char)

    def deleteText(self, k: int) -> int:
        # Delete up to k characters from the left stack
        count = 0
        while k > 0 and self.left_stack:
            self.left_stack.pop()
            count += 1
            k -= 1
        return count

    def cursorLeft(self, k: int) -> str:
        # Move k characters from left stack to right stack
        while k > 0 and self.left_stack:
            self.right_stack.append(self.left_stack.pop())
            k -= 1
        return self._get_last_10()

    def cursorRight(self, k: int) -> str:
        # Move k characters from right stack to left stack
        while k > 0 and self.right_stack:
            self.left_stack.append(self.right_stack.pop())
            k -= 1
        return self._get_last_10()

    def _get_last_10(self) -> str:
        # Helper to return the last min(10, len) characters from left stack
        # Since slicing handles out of bounds gracefully, we can just slice.
        # The slice [-10:] takes the last 10 elements. If fewer than 10, it takes all.
        return "".join(self.left_stack[-10:])

# Your TextEditor object will be instantiated and called as such:
# obj = TextEditor()
# obj.addText(text)
# param_2 = obj.deleteText(k)
# param_3 = obj.cursorLeft(k)
# param_4 = obj.cursorRight(k)
# @lc code=end

File: 2296.design-a-text-editor.cpp
#
# @lc app=leetcode id=2296 lang=cpp
#
# [2296] Design a Text Editor
#

# @lc code=start
class TextEditor {
private:
    string L,R;
public:
    TextEditor() {
        
    }
    
    void addText(string text) {
        L.append(text);
    }
    
    int deleteText(int k) {
        int cnt = min(k,(int)L.size());
        L.resize(L.size()-cnt);
        return cnt;
    }
    
    string cursorLeft(int k) {
        int mv = min(k,(int)L.size());
        while(mv--) {
            R.push_back(L.back());
            L.pop_back();
        }
        int len = L.size();
        int start = max(0,len-10);
        return L.substr(start);
    }
    
    string cursorRight(int k) {
        int mv = min(k,(int)R.size());
        while(mv--) {
            L.push_back(R.back());
            R.pop_back();
        }
        int len = L.size();
        int start = max(0,len-10);
        return L.substr(start);
    }
};

/**
 * Your TextEditor object will be instantiated and called as such:
 * TextEditor* obj = new TextEditor();
 * obj->addText(text);
 * int param_2 = obj->deleteText(k);
 * string param_3 = obj->cursorLeft(k);
 * string param_4 = obj->cursorRight(k);
 */
# @lc code=end